onnxruntime/core/providers/cpu/reduction/reduction_ops.h

```diff
@@ -132,6 +180,86 @@ class ReduceAggregatorSum : public ReduceAggregator<T, TVAL> {
   inline TVAL aggall(const T* from_data) {
     return Eigen::Map<const Eigen::Matrix<T, Eigen::Dynamic, 1>>(from_data, this->N_).sum();
   }
+
+  // Fast reduction
+  static inline FastReduceKind fast_reduce() { return FastReduceKindValues::KR | FastReduceKindValues::RK | FastReduceKindValues::KRK; }
+
+  static void FastReduceKR(const Tensor& input, const std::vector<int64_t>& fast_shape,
+                           Tensor& output, concurrency::ThreadPool* tp) {
+    ORT_ENFORCE(fast_shape.size() == 2, "Only works on matrices with two dimensions.");
+    ORT_ENFORCE(fast_shape[0] == output.Shape().Size(), "Output size mismatch.");
+    // TODO: use MLAS or BLAS
+    if (fast_shape[0] >= 4) {
+      const T* data = input.Data<T>();
+      T* out = output.MutableData<T>();
+      int64_t stridei = fast_shape[1];
+      concurrency::ThreadPool::TryBatchParallelFor(
+          tp,
+          SafeInt<int32_t>(fast_shape[0]),
+          [data, stridei, out](ptrdiff_t j) {
+            out[j] = ConstEigenVectorArrayMap<T>(data + j * stridei, stridei).sum();
+          },
+          0);
+    } else {
+      math::RowwiseSum<T, CPUMathUtil>((int)fast_shape[0], (int)fast_shape[1], input.Data<T>(), output.MutableData<T>(), nullptr);
+    }
+  }
+
```

Comment:
There seems to be a fair bit of duplicated code across the different reduction types. Can we just plugin a delegate to add the Eigen call (e.g. the sum()/maxCoeff()) and re-use more?